File: crates/core/src/settings/hotkeys.rs
//! Hotkeys settings (global Hotkeys tab): a remappable keymap of named actions.
//!
//! The action set and their **default** bindings are code-defined here — the single source of truth.
//! The durable document stores only the user's deviations: a remap ([`Binding`]) or a disable
//! (`None`). An action with no stored override uses its default, so "Reset all to defaults" is simply
//! clearing the overrides, and a future build that changes a default reaches every user who has not
//! overridden that action.
//!
//! Each action has a [`HotkeyScope`]; the same key may bind different actions in different scopes
//! (e.g. "previous project" in the sidebar and "previous process" in the terminal), so a conflict is
//! only ever within one scope. macOS `⌘`/`⌥` from Solo's reference are remapped to Ctrl/Alt for Linux.
//! Key tokens follow the web `KeyboardEvent.key` convention so the frontend matches a real key event
//! directly.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

/// The context a hotkey is active in. Bindings only conflict within the same scope.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HotkeyScope {
    General,
    Sidebar,
    Terminal,
}

/// A named, remappable action. The closed set is the single source the settings document and the
/// frontend keyboard handler iterate.
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HotkeyAction {
    // General — app-wide actions, palettes, and system shortcuts.
    OpenCommandPalette,
    QuickActions,
    QuickJump,
    NewAgentOrTerminal,
    OpenSettings,
    OpenTerminalSearch,
    CloseAgentOrTerminal,
    // Sidebar — navigation.
    NextProjectGroup,
    PrevProjectGroup,
    NextSection,
    PrevSection,
    JumpToAgents,
    JumpToCommands,
    JumpToTerminals,
    CollapseOrSection,
    JumpToParentProject,
    ExpandProject,
    RestartSelection,
    // Terminal — active while the terminal is focused.
    PreviousProcess,
    NextProcess,
    IncreaseTerminalFontSize,
    DecreaseTerminalFontSize,
}

/// A key chord: the modifier flags plus the main key (a `KeyboardEvent.key` token, e.g. `K`,
/// `ArrowDown`, `=`). Stored as data, compared by value for conflict detection.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct Binding {
    pub ctrl: bool,
    pub alt: bool,
    pub shift: bool,
    #[serde(rename = "super")]
    pub super_key: bool,
    pub key: String,
}

impl Binding {
    /// A `Ctrl`-modified chord (the Linux remap of a Solo `⌘` shortcut).
    fn ctrl(key: &str) -> Self {
        Self {
            ctrl: true,
            alt: false,
            shift: false,
            super_key: false,
            key: key.to_string(),
        }
    }

    /// An `Alt`-modified chord (the Linux remap of a Solo `⌥`/Option shortcut).
    fn alt(key: &str) -> Self {
        Self {
            ctrl: false,
            alt: true,
            shift: false,
            super_key: false,
            key: key.to_string(),
        }
    }

    /// An unmodified key (e.g. a single-letter sidebar shortcut).
    fn plain(key: &str) -> Self {
        Self {
            ctrl: false,
            alt: false,
            shift: false,
            super_key: false,
            key: key.to_string(),
        }
    }
}

impl HotkeyAction {
    /// Every action, in display order — the single list the document and UI iterate.
    pub const ALL: [HotkeyAction; 22] = [
        HotkeyAction::OpenCommandPalette,
        HotkeyAction::QuickActions,
        HotkeyAction::QuickJump,
        HotkeyAction::NewAgentOrTerminal,
        HotkeyAction::OpenSettings,
        HotkeyAction::OpenTerminalSearch,
        HotkeyAction::CloseAgentOrTerminal,
        HotkeyAction::NextProjectGroup,
        HotkeyAction::PrevProjectGroup,
        HotkeyAction::NextSection,
        HotkeyAction::PrevSection,
        HotkeyAction::JumpToAgents,
        HotkeyAction::JumpToCommands,
        HotkeyAction::JumpToTerminals,
        HotkeyAction::CollapseOrSection,
        HotkeyAction::JumpToParentProject,
        HotkeyAction::ExpandProject,
        HotkeyAction::RestartSelection,
        HotkeyAction::PreviousProcess,
        HotkeyAction::NextProcess,
        HotkeyAction::IncreaseTerminalFontSize,
        HotkeyAction::DecreaseTerminalFontSize,
    ];

    /// The scope this action is active in.
    pub fn scope(self) -> HotkeyScope {
        use HotkeyAction::*;
        match self {
            OpenCommandPalette | QuickActions | QuickJump | NewAgentOrTerminal | OpenSettings
            | OpenTerminalSearch | CloseAgentOrTerminal => HotkeyScope::General,
            NextProjectGroup | PrevProjectGroup | NextSection | PrevSection | JumpToAgents
            | JumpToCommands | JumpToTerminals | CollapseOrSection | JumpToParentProject
            | ExpandProject | RestartSelection => HotkeyScope::Sidebar,
            PreviousProcess | NextProcess | IncreaseTerminalFontSize | DecreaseTerminalFontSize => {
                HotkeyScope::Terminal
            }
        }
    }

    /// The code-defined default binding (Solo's `⌘`/`⌥` reference remapped to Ctrl/Alt for Linux).
    pub fn default_binding(self) -> Binding {
        use HotkeyAction::*;
        match self {
            OpenCommandPalette => Binding::ctrl("K"),
            QuickActions => Binding::ctrl("P"),
            QuickJump => Binding::ctrl("E"),
            NewAgentOrTerminal => Binding::ctrl("T"),
            OpenSettings => Binding::ctrl(","),
            OpenTerminalSearch => Binding::ctrl("F"),
            CloseAgentOrTerminal => Binding::ctrl("W"),
            NextProjectGroup => Binding::ctrl("ArrowDown"),
            PrevProjectGroup => Binding::ctrl("ArrowUp"),
            NextSection => Binding::alt("ArrowDown"),
            PrevSection => Binding::alt("ArrowUp"),
            JumpToAgents => Binding::alt("A"),
            JumpToCommands => Binding::alt("C"),
            JumpToTerminals => Binding::alt("T"),
            CollapseOrSection => Binding::plain("ArrowLeft"),
            JumpToParentProject => Binding::ctrl("ArrowLeft"),
            ExpandProject => Binding::plain("ArrowRight"),
            RestartSelection => Binding::plain("R"),
            PreviousProcess => Binding::ctrl("ArrowUp"),
            NextProcess => Binding::ctrl("ArrowDown"),
            IncreaseTerminalFontSize => Binding::ctrl("="),
            DecreaseTerminalFontSize => Binding::ctrl("-"),
        }
    }
}

/// One action's effective state in the read model the UI renders.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct HotkeyBindingView {
    pub action: HotkeyAction,
    pub scope: HotkeyScope,
    /// The effective binding, or `None` when the action is disabled.
    pub binding: Option<Binding>,
    /// Whether the effective binding is the code default (no user override).
    pub is_default: bool,
    /// Whether this binding collides with another action in the same scope (`conflicts`) — the
    /// rows the UI flags. Carried in the view so the frontend never re-derives the rule.
    pub conflict: bool,
}

/// The Hotkeys tab document. Stores only deviations from the defaults: `Some(binding)` is a remap,
/// `None` is a disabled action; an action absent from the map uses its code default. So the persisted
/// record stays small and a default change reaches everyone who has not overridden the action.
#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct Hotkeys {
    overrides: BTreeMap<HotkeyAction, Option<Binding>>,
}

impl Hotkeys {
    /// The effective binding for `action` — the override if one is stored (a remap, or `None` when
    /// disabled), otherwise the code default.
    pub fn binding(&self, action: HotkeyAction) -> Option<Binding> {
        match self.overrides.get(&action) {
            Some(over) => over.clone(),
            None => Some(action.default_binding()),
        }
    }

    /// Remaps `action` to `binding` (an override).
    pub fn remap(&mut self, action: HotkeyAction, binding: Binding) {
        self.overrides.insert(action, Some(binding));
    }

    /// Disables `action` (hover-and-press-x): it keeps no binding until reset.
    pub fn disable(&mut self, action: HotkeyAction) {
        self.overrides.insert(action, None);
    }

    /// Resets `action` to its code default by dropping any override.
    pub fn reset(&mut self, action: HotkeyAction) {
        self.overrides.remove(&action);
    }

    /// Resets every action to its code default ("Reset all to defaults").
    pub fn reset_all(&mut self) {
        self.overrides.clear();
    }

    /// The full keymap read model — every action with its scope, effective binding, and whether it
    /// is still the default.
    pub fn view(&self) -> Vec<HotkeyBindingView> {
        let conflicts: std::collections::BTreeSet<HotkeyAction> =
            self.conflicts().into_iter().collect();
        HotkeyAction::ALL
            .into_iter()
            .map(|action| HotkeyBindingView {
                action,
                scope: action.scope(),
                binding: self.binding(action),
                is_default: !self.overrides.contains_key(&action),
                conflict: conflicts.contains(&action),
            })
            .collect()
    }

    /// Actions whose effective binding collides with another action **in the same scope** — the
    /// conflicts the UI flags. A shared key across different scopes is never a conflict.
    pub fn conflicts(&self) -> Vec<HotkeyAction> {
        let mut clashing = Vec::new();
        let bound: Vec<_> = HotkeyAction::ALL
            .into_iter()
            .filter_map(|a| self.binding(a).map(|b| (a, b)))
            .collect();
        for (action, binding) in &bound {
            let collides = bound.iter().any(|(other, other_binding)| {
                other != action && other.scope() == action.scope() && other_binding == binding
            });
            if collides {
                clashing.push(*action);
            }
        }
        clashing
    }
}

#[cfg(test)]
#[path = "hotkeys_tests.rs"]
mod tests;

File: crates/app/ui/src/lib/hotkeys.ts
// The Hotkeys tab's display data and the chord <-> KeyboardEvent conversion. The action set,
// scopes, and default bindings live in the core (the single source); this maps each closed enum
// to its human label and renders / parses a chord the way the core stores it (uppercase letters,
// `KeyboardEvent.key` tokens) so a captured chord round-trips and the live handler matches a real
// key event.

import type { Binding, HotkeyAction, HotkeyScope } from "@/domain";

export const HOTKEY_SCOPE_ORDER: HotkeyScope[] = ["general", "sidebar", "terminal"];

export const HOTKEY_SCOPE_LABELS: Record<HotkeyScope, string> = {
  general: "General",
  sidebar: "Sidebar",
  terminal: "Terminal",
};

export const HOTKEY_SCOPE_DESCRIPTIONS: Record<HotkeyScope, string> = {
  general: "App-wide actions, palettes, and system shortcuts.",
  sidebar: "Navigating the project tree.",
  terminal: "Active while the terminal is focused.",
};

// The human label for each action — one source the panel and search read.
export const HOTKEY_ACTION_LABELS: Record<HotkeyAction, string> = {
  open_command_palette: "Open command palette",
  quick_actions: "Quick actions",
  quick_jump: "Quick jump",
  new_agent_or_terminal: "New agent or terminal",
  open_settings: "Open settings",
  open_terminal_search: "Open terminal search",
  close_agent_or_terminal: "Close agent or terminal",
  next_project_group: "Next project group",
  prev_project_group: "Previous project group",
  next_section: "Next section",
  prev_section: "Previous section",
  jump_to_agents: "Jump to Agents",
  jump_to_commands: "Jump to Commands",
  jump_to_terminals: "Jump to Terminals",
  collapse_or_section: "Collapse / go to section",
  jump_to_parent_project: "Jump to parent project",
  expand_project: "Expand project",
  restart_selection: "Restart",
  previous_process: "Previous process",
  next_process: "Next process",
  increase_terminal_font_size: "Increase terminal font size",
  decrease_terminal_font_size: "Decrease terminal font size",
};

// The keys that are modifiers themselves — a chord is only complete once a non-modifier key is
// pressed, so a capture ignores these.
const MODIFIER_KEYS = new Set(["Control", "Alt", "Shift", "Meta"]);

// How a stored key token renders to the eye (arrows as glyphs, space named, letters as-is).
const KEY_GLYPHS: Record<string, string> = {
  ArrowUp: "↑",
  ArrowDown: "↓",
  ArrowLeft: "←",
  ArrowRight: "→",
  " ": "Space",
  Escape: "Esc",
};

// The ordered modifier tokens of a chord, then its key — e.g. ["Ctrl", "K"] or ["Alt", "↑"].
export function formatChord(binding: Binding): string[] {
  const tokens: string[] = [];
  if (binding.ctrl) tokens.push("Ctrl");
  if (binding.alt) tokens.push("Alt");
  if (binding.shift) tokens.push("Shift");
  if (binding.super) tokens.push("Super");
  tokens.push(KEY_GLYPHS[binding.key] ?? binding.key);
  return tokens;
}

// Build the chord a key event represents, the way the core stores it. Returns null while only a
// modifier is held (the capture waits for a real key). Single letters are uppercased so a press
// matches the core's "K"-style defaults without depending on the Shift state.
export function bindingFromEvent(event: KeyboardEvent): Binding | null {
  if (MODIFIER_KEYS.has(event.key)) return null;
  const key =
    event.key.length === 1 && /[a-z]/i.test(event.key) ? event.key.toUpperCase() : event.key;
  return {
    ctrl: event.ctrlKey,
    alt: event.altKey,
    shift: event.shiftKey,
    super: event.metaKey,
    key,
  };
}

// Whether a chord carries a command modifier (Ctrl/Alt/Super). Shift alone is still typing, so
// it does not count. App-wide shortcuts with a command modifier fire even while a text field is
// focused; bare-key shortcuts must yield to typing, so the live handler uses this to decide.
export function hasCommandModifier(binding: Binding): boolean {
  return binding.ctrl || binding.alt || binding.super;
}

export function bindingsEqual(a: Binding, b: Binding): boolean {
  return (
    a.ctrl === b.ctrl &&
    a.alt === b.alt &&
    a.shift === b.shift &&
    a.super === b.super &&
    a.key === b.key
  );
}

// True when the event originates in a text-editing surface (input, textarea, or
// contenteditable). A bare-key shortcut must yield to typing there; command-modifier
// shortcuts (Ctrl/Alt/Super) still fire everywhere, as in a native app.
export function isEditableTarget(target: EventTarget | null): boolean {
  if (!(target instanceof HTMLElement)) return false;
  return target.tagName === "INPUT" || target.tagName === "TEXTAREA" || target.isContentEditable;
}
